Restrict HubProtocol on Server + HubOptions<THub>

mikaelm12 opened the pull request:
Adding the ability to restrict the HubProtocol on the server
Issue: https://github.com/aspnet/SignalR/issues/854

davidfowl commented:
I think this is way too global (I was going to update the issue but this is fine). It needs to be per hub end point not per signalr instance. We might need HubOptions<THub> to support this properly.

anurse commented on src/Microsoft.AspNetCore.SignalR.Common/DefaultProtocols.cs:
```diff
@@ -0,0 +1,11 @@
+{
+    public static class DefaultProtocols
+    {
+        public static readonly string JSONProtocol = "json";
```
Use `Json`. We generally treat it like a "Word" in our code, even though it is technically an acronym ;)

mikaelm12 replied:
I made it all lowercase to match the name property on the JsonHubProtocol https://github.com/aspnet/SignalR/blob/dev/src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/JsonHubProtocol.cs#L31
I can change that too if we want.

davidfowl replied:
The only default protocol we have is JSON really. MSGpack is an addon.

mikaelm12 replied:
Okay, I can remove that from here. Was following what the issue laid out 😄

davidfowl commented:
We'll remove the HubOptions, in the AddSignalR overload and instead make them `HubOptions<THub>`. We'll revisit global options later.

`IHubProtocol.GetProtocol` now takes 3 arguments:

```C#
GetProtocol(IReadOnlyList<string> supportedProtocols,
string requestedProtocols,
HubConnectionContext context); 
```

We also need to add an extension method AddHubOptions on ISignalRBuilder to enable configuring this:

```C#
services.AddSignalR()
              .AddHubOptions<Chat>(o =>
              {
                   o.KeepAliveTimeout = TimeSpan.FromSeconds(5);
                   o.SupportedProtocols.Add("msgpack");
              });
```

mikaelm12 commented:
Okay, replaced `HubOptions` with `HubOptions<THub>`. HubOptions are no longer global. This probably could have logically been split into two PRs but oh well. 
The internals are quite different now btw

davidfowl commented on src/Microsoft.AspNetCore.SignalR.Core/HubOptions.cs:
```diff
@@ -24,5 +25,7 @@ public class HubOptions
         /// </remarks>
         public TimeSpan KeepAliveInterval { get; set; } = DefaultKeepAliveInterval;
+
+        public List<string> SupportedProtocols { get; set; } = new List<string> { DefaultProtocols.JSONProtocol, DefaultProtocols.MsgPackProtocol};
```
I was waiting for this to come up. I think by default you support all available transports. That is, not specified means all. There also needs to be validation like what if you clear all? What does that mean?

BrennanConroy replied:
Could we somehow grab all available transports when HubOptions is created and populate the list before running the users action on it?

davidfowl replied:
We have nested generics so I'm not sure how well it will work. You'd need to get all the configured IHubProtocol implementations inside of an `IConfigureOptions<HubOptions<THub>>` which makes it a bit difficult as there's no explicit call to register a hub.

anurse replied:
Other fun things: MsgPack is in a completely separate assembly that isn't available in `AspNetCore.All` currently.

mikaelm12 commented on samples/SocketsSample/Startup.cs:
```diff
@@ -20,11 +20,25 @@ public void ConfigureServices(IServiceCollection services)
+            })
+            .AddHubOptions<DynamicChat>(options =>
             {
-                // Faster pings for testing
```
Don't know why I removed this. I'll bring it back

davidfowl commented on src/Microsoft.AspNetCore.SignalR/SignalRDependencyInjectionExtensions.cs:
```diff
@@ -15,10 +18,49 @@ public static ISignalRBuilder AddSignalR(this IServiceCollection services)
                 .AddJsonProtocol();
         }
 
+        public static ISignalRBuilder AddGlobalHubOptions(this ISignalRBuilder signalRBuilder, Action<HubOptions> options)
```
This should be an overload of AddSignalR. There should be no global options call.

mikaelm12 replied:
Makes sense

davidfowl commented on src/Microsoft.AspNetCore.SignalR/SignalRDependencyInjectionExtensions.cs:
```diff
@@ -15,10 +18,49 @@ public static ISignalRBuilder AddSignalR(this IServiceCollection services)
         }
     }
+
+    public class HubOptionsSetup : IConfigureOptions<HubOptions> 
```
Put these in separate files.

mikaelm12 commented:
Okay, this is working now and is ready for review

davidfowl commented on samples/SocketsSample/Startup.cs:
```diff
@@ -20,11 +20,11 @@ public void ConfigureServices(IServiceCollection services)
                 // Faster pings for testing
                 options.KeepAliveInterval = TimeSpan.FromSeconds(5);
             })
+
```
nit: remove space

davidfowl commented on src/Microsoft.AspNetCore.SignalR.Core/HubOptionsSetup.cs:
```diff
@@ -0,0 +1,33 @@
+
+        public void Configure(HubOptions options)
+        {
+            if( options.SupportedProtocols == null)
```
nit: formatting

davidfowl commented on src/Microsoft.AspNetCore.SignalR.Core/HubOptionsSetup`T.cs:
```diff
@@ -0,0 +1,31 @@
+
+        public void Configure(HubOptions<THub> options)
+        {
+            if(_hubOptions.SupportedProtocols == null)
```
Can this ever be null?

mikaelm12 replied:
It shouldn't be. I'll remove this check

davidfowl commented on src/Microsoft.AspNetCore.SignalR.Core/Internal/DefaultHubProtocolResolver.cs:
```diff
@@ -33,11 +31,11 @@ public DefaultHubProtocolResolver(IOptions<HubOptions> options, IEnumerable<IHub
             protocolName = protocolName ?? throw new ArgumentNullException(nameof(protocolName));
 
-            if (_availableProtocols.TryGetValue(protocolName, out var protocol))
+            if (_availableProtocols.TryGetValue(protocolName, out var protocol) && (supportedProtocols ==  null || supportedProtocols.Contains(protocolName)))
```
What about casing? Do the case rules match the dictionary? It should be insensitive.

davidfowl commented on src/Microsoft.AspNetCore.SignalR.Core/SignalRDependencyInjectionExtensions.cs:
```diff
@@ -4,6 +4,7 @@
 using Microsoft.AspNetCore.SignalR;
 using Microsoft.AspNetCore.SignalR.Core;
 using Microsoft.AspNetCore.SignalR.Internal;
+using Microsoft.Extensions.Options;
```
Revert this file?

davidfowl commented on src/Microsoft.AspNetCore.SignalR/HubOptionsDependencyInjectionExtensions.cs:
```diff
@@ -0,0 +1,20 @@
+{
+    public static class HubOptionsDependencyInjectionExtensions
+    {
+
```
nit: remove empty space.

davidfowl commented on src/Microsoft.AspNetCore.SignalR/SignalRDependencyInjectionExtensions.cs:
```diff
@@ -3,22 +3,25 @@
+
+        public static ISignalRBuilder AddSignalR(this IServiceCollection services, Action<HubOptions> options)
+        {
+            services.AddSingleton<IConfigureOptions<HubOptions>, HubOptionsSetup>()
```
This should be in the parameterless AddSignalR

davidfowl commented on test/Microsoft.AspNetCore.SignalR.Tests/HubEndpointTests.cs:
```diff
@@ -274,7 +274,7 @@ public class HubEndpointTests
             var serviceProvider = HubEndPointTestUtils.CreateServiceProvider(services =>
             {
-                services.Configure<HubOptions>(hubOptions =>
+                services.Configure<HubOptions<SimpleHub>>(hubOptions =>
```
Why isn't this just the same as before? why did you change it?

davidfowl commented on test/Microsoft.AspNetCore.SignalR.Tests/HubEndpointTests.cs:
```diff
@@ -1325,10 +1325,15 @@ public void HubsCannotHaveOverloadedMethods()
-            var serviceProvider = HubEndPointTestUtils.CreateServiceProvider();
+            var serviceProvider = HubEndPointTestUtils.CreateServiceProvider(services =>
+            {
+                services.AddSignalR(options =>
```
Why did you add this?

davidfowl commented on test/Microsoft.AspNetCore.SignalR.Tests/HubEndpointTests.cs:
```diff
@@ -1673,7 +1678,7 @@ public static IEnumerable<object[]> StreamingMethodAndHubProtocols
         {
             var serviceProvider = HubEndPointTestUtils.CreateServiceProvider(services =>
-                services.Configure<HubOptions>(options =>
+                services.Configure<HubOptions<MethodHub>>(options =>
```
Why did you change these?

davidfowl commented on test/Microsoft.AspNetCore.SignalR.Tests/Internal/DefaultHubProtocolResolverTests.cs:
```diff
@@ -22,18 +22,48 @@ public class DefaultHubProtocolResolverTests
             new MessagePackHubProtocol()
         };
 
+        private static readonly List<string> AllProtocolNames = new List<string> { "json", "messagepack" };
```
Why do you have this?

mikaelm12 commented:
Okay. The tests weren't actually testing the correct thing before. I fixed them to ensure that the global hub options are being configured.

I changed the HubOptions<THub> class values to be null by default to add logic to determine if we should use the global options. We have to do this because configure on HubOptionsSetup<THub> doesn't run unless we explicitly call `addHubOptions` for a specific THub.

BrennanConroy commented on samples/SocketsSample/Startup.cs:
```diff
@@ -20,8 +20,7 @@ public void ConfigureServices(IServiceCollection services)
 
-            services.AddSignalR(options =>
-            {
+            services.AddSignalR(options => {
```
revert

BrennanConroy commented on src/Microsoft.AspNetCore.SignalR.Core/HubEndPoint.cs:
```diff
@@ -32,16 +34,30 @@ public class HubEndPoint<THub> where THub : Hub
         public async Task OnConnectedAsync(ConnectionContext connection)
         {
-            var connectionContext = new HubConnectionContext(connection, _hubOptions.KeepAliveInterval, _loggerFactory);
+            var keepAlive = _hubOptions.KeepAliveInterval ?? _globalHubOptions.KeepAliveInterval;
```
Add a comment saying `HubOptions<T>` is preferred over `HubOptions`

BrennanConroy commented on src/Microsoft.AspNetCore.SignalR.Core/HubOptionsSetup.cs:
```diff
@@ -0,0 +1,33 @@
+
+        public HubOptionsSetup(IEnumerable<IHubProtocol> protocols)
+        {
+            foreach (IHubProtocol hubProtocol in protocols)
```
var

BrennanConroy commented on src/Microsoft.AspNetCore.SignalR.Core/Internal/DefaultHubProtocolResolver.cs:
```diff
@@ -33,11 +32,11 @@ public DefaultHubProtocolResolver(IOptions<HubOptions> options, IEnumerable<IHub
             protocolName = protocolName ?? throw new ArgumentNullException(nameof(protocolName));
 
-            if (_availableProtocols.TryGetValue(protocolName, out var protocol))
+            if (_availableProtocols.TryGetValue(protocolName, out var protocol) && (supportedProtocols ==  null || supportedProtocols.Contains(protocolName, StringComparer.OrdinalIgnoreCase)))
```
nit: remove extra space

BrennanConroy commented on src/Microsoft.AspNetCore.SignalR.Core/Internal/IHubProtocolResolver.cs:
```diff
@@ -2,11 +2,12 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using Microsoft.AspNetCore.SignalR.Internal.Protocol;
+using System.Collections.Generic;
```
nit: sort

davidfowl commented on test/Microsoft.AspNetCore.SignalR.Tests/Internal/DefaultHubProtocolResolverTests.cs:
```diff
@@ -22,18 +22,47 @@ public class DefaultHubProtocolResolverTests
             new MessagePackHubProtocol()
         };
 
+        private static readonly List<string> AllProtocolNames = new List<string> { "json", "messagepack" };
```
Why are you hardcode this?

davidfowl replied:
move to top of the file?

JamesNK commented on src/Microsoft.AspNetCore.SignalR.Core/HubOptions`T.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    public class HubOptions<THub> : HubOptions where THub : Hub
+    {
+        public new TimeSpan? NegotiateTimeout { get; set; } = null;
```
It feels really weird to me that these properties are defined again and are using new to hide the original properties. Is it because you want NegotiateTimeout and KeepAliveInterval to be nullable here? What scenario is this for?

davidfowl replied:
Yea, maybe they should just be null in the base class.

mikaelm12 replied:
Good question. 
The properties are null in this class and not the base class because we want the default values from the base class, but we need to be able to tell if a specific `HubOptions<THub>` has been configured (aka the options are not null) 
It allows us to differentiate between the scenarios in which there are only global options defined, global and local so we know which settings to enforce. 

That being said, yeah we could just make them null in the base class but I do like the idea of default values.

davidfowl replied:
We can just have defaults come from the IConfigureOptions<HubOptions> instead of being on the HubOptions class. It's a bit cleaner.

davidfowl commented on src/Microsoft.AspNetCore.SignalR.Core/HubOptionsSetup.cs:
```diff
@@ -26,6 +27,18 @@ public void Configure(HubOptions options)
+                options.KeepAliveInterval = TimeSpan.FromSeconds(15);
+            }
+
+            if(options.NegotiateTimeout == null)
```
nit: formatting

davidfowl commented on src/Microsoft.AspNetCore.SignalR.Core/HubEndPoint.cs:
```diff
@@ -32,16 +34,31 @@ public class HubEndPoint<THub> where THub : Hub
+            var connectionContext = new HubConnectionContext(connection, keepAlive.Value, _loggerFactory);
 
-            if (!await connectionContext.NegotiateAsync(_hubOptions.NegotiateTimeout, _protocolResolver, _userIdProvider))
+            if (!await connectionContext.NegotiateAsync(negotiateTimeout.Value, supportedProtocols, _protocolResolver, _userIdProvider))
```
This can throw if negotiateTimeout is null...

mikaelm12 replied:
It shouldn't be null since it gets set in the HubOptionsSetup

davidfowl replied:
What if I explicitly set it to null?

mikaelm12 commented on src/Microsoft.AspNetCore.SignalR.Core/HubEndPoint.cs:
```diff
@@ -32,16 +34,31 @@ public class HubEndPoint<THub> where THub : Hub
-            var connectionContext = new HubConnectionContext(connection, _hubOptions.KeepAliveInterval, _loggerFactory);
+            // We check to see if HubOptions<THub> are set because those take precedence over global hub options.
+            var keepAlive = _hubOptions.KeepAliveInterval ?? _globalHubOptions.KeepAliveInterval;
+            var negotiateTimeout = _hubOptions.NegotiateTimeout ?? _globalHubOptions.NegotiateTimeout;
```
I think this makes sense.

JamesNK commented on src/Microsoft.AspNetCore.SignalR.Core/HubEndPoint.cs:
```diff
@@ -32,16 +34,32 @@ public class HubEndPoint<THub> where THub : Hub
+            var keepAlive = _hubOptions.KeepAliveInterval ?? _globalHubOptions.KeepAliveInterval ?? HubOptionsSetup.DefaultKeepAliveInterval;
+            var negotiateTimeout = _hubOptions.NegotiateTimeout ?? _globalHubOptions.NegotiateTimeout ?? HubOptionsSetup.DefaultNegotiateTimeout;
+            var supportedProtocols = _hubOptions.SupportedProtocols;
+            if (supportedProtocols == null && _globalHubOptions.SupportedProtocols != null)
```
You could replace this if statement with a null coalesce on the line above. It doesn't matter if `_globalHubOptions.SupportedProtocols` protocols is null based on the check line 55.

anurse commented on src/Microsoft.AspNetCore.SignalR.Core/HubOptions.cs:
```diff
@@ -2,27 +2,16 @@
+        public TimeSpan? NegotiateTimeout { get; set; } = null;
 
-        public TimeSpan NegotiateTimeout { get; set; } = TimeSpan.FromSeconds(5);
+        public TimeSpan? KeepAliveInterval { get; set; } = null;
```
Bring the docs back, and add a note about `null` indicating the default interval (15 seconds).

JamesNK replied:
The comment is a bit more complicated than that because `null` could indicate the default, or it could indicate that the global option should be used

anurse replied:
Ugh, yeah. Ok, save it for now, we'll need to doc it at some point though.

anurse commented on src/Microsoft.AspNetCore.SignalR.Core/HubOptions.cs:
```diff
@@ -2,27 +2,16 @@
-        /// functionality because of the polling mechanism. 
-        /// </remarks>
-        public TimeSpan KeepAliveInterval { get; set; } = DefaultKeepAliveInterval;
+        public List<string> SupportedProtocols { get; set; } = null;
```
Make this `IList<string>`

We're saying `null` is "any protocol" right? Let's throw a doc comment on this explaining it as well.

anurse commented on src/Microsoft.AspNetCore.SignalR.Core/Internal/IHubProtocolResolver.cs:
```diff
@@ -1,12 +1,13 @@
     public interface IHubProtocolResolver
     {
-        IHubProtocol GetProtocol(string protocolName, HubConnectionContext connection);
+        IHubProtocol GetProtocol(string protocolName, List<string> supportedProtocols, HubConnectionContext connection);
```
`List<T>` in a public API!

![https://media.giphy.com/media/sFTWiBKYYWKVa/giphy.gif](https://media.giphy.com/media/sFTWiBKYYWKVa/giphy.gif)

Use `IReadOnlyList<string>`. The method only needs to be able to read from the list, and doesn't care what kind of list you give it.

mikaelm12 replied:
This should also just be an `IList<string>`, right? It get passed the supportedProtocol list from the HubEndPoint which is an IList

JamesNK replied:
Should it not be `List<T>` even if it is in the internal namespace? I wasn't sure about that when I looked at the code.

anurse replied:
In this case, the API is likely to be made properly public at some point. Also, "pubternal" (public in the Internal namespace) types should still try to keep to our public API standards.
